### ISSUE TITLE
Java Conversion of Event Fields should be Lazy

### ISSUE THREAD

**Issue by original-brownbear** [elastic/logstash#7803]
I think we have huge optimization potential when it comes to setting and retrieving data from an `Event`.

Currently this example in Ruby:

```ruby
 e = LogStash::Event.new({
"foo" => "bar", "bar" => 1, "baz" => 1.0, TIMESTAMP => "2015-05-28T23:02:05.350Z"
})
 e.get("foo")
```

causes the following things to happen under the hood:

#### Instantiation
* Convert `RubyHash`  to `ConvertedMap`
* Create a BiValue for every entry in the `ConvertedMap`
* Create `Event` object with a bunch of additional `HashMap<>` etc.

#### Get
* Turn the key "foo" from a `RubyString` into a `String`
* Lookup on `Event` which creates an entry in the `PathCache`
   * and some more objects in `Accessor`

-------------------

This is just wrong in my opinion and an extreme overhead memory wise (not that in the apache logs benchmark example you can easily attribute `50%+` of the GC activity to the above redundant steps).

IMO:

* No Java `Event` should be created unless we have a Java access to it or want to serialize it
* all operations on the `RubyEvent` should stay strictly in JRuby class space without any unnecessary instantiations/conversions
* We should only switch over to the current code once the `RubyEvent` has seen a Java type access

**Comment by original-brownbear** [comment]
@jordansissel @suyograo wdyt? This is probably **the** optimization when it comes to GC/Memory and a relatively low hanging fruit as well (I mean the code to just do the field access in Ruby must be around in the repo somewhere still and probably just needs minor changes if any).
+ Even the Java plugins + PQ should get faster with this (though not by as much as pure Ruby access since lazy initialization would way decreases the lifetime of the converted Java objects

**Comment by jordansissel** [comment]
I agree with the premise (lazy object creation), but I am not sure the operating environment will enable this kind of optimization because most users are probably going to execute at least one filter that's in Java (date, for example).

This would be an easier "yes" if you could provide data to show the improvement.

**Comment by original-brownbear** [comment]
@jordansissel 

> because most users are probably going to execute at least one filter that's in Java (date, for example).

We could simply make it lazy by field and simply only convert the "tree" of data that was touched by Java?

> This would be an easier "yes" if you could provide data to show the improvement.

Well I can see a few things from data running the Apache dataset with `-w 4`:

* more RAM currently means faster (this holds true for up to `4GB Xmx` on my box)
   * This is not cool, considering that the test dataset is just ~2GB.
* 50% of the garbage created (at least, likely more but allocation tracking in a profiler is tricky) goes towards all this conversion logic
* Our stacks in the `get()` calls in some cases are so deep that the JIT gives up with the default depth of 9
* we burn a ton of cycles for method calls already (in part as a result of the above), we need to simplify and this is a very straightforward spot for it

=> This will def have a very visible improvement in memory use (with all the attached benefits of wasting fewer cycles to GC, having a higher CPU cache hit rate and all that :))
=> Stacks get flatter => JIT gets happier and further optimizations become clearer

**Comment by original-brownbear** [comment]
@jordansissel actually the fact that we got a visible change in this PR is hard evidence here imo:

See https://github.com/elastic/logstash/pull/7603#issuecomment-313525142 (removing just a few StringBiValue got us ~5-10% with 2 threads, more with 4 workers)

**Comment by original-brownbear** [comment]
Also note that the whole `BiValue` logic is redundant now anyway as explained in https://github.com/elastic/logstash/issues/7712. So basically this would just be a way to also avoid needlessly converting hashes and keys. Values we shouldn't be converting in the first place imo.

**Comment by original-brownbear** [comment]
Or just take a look at this file to see the GC impact.

Config:

```
input {
  stdin { }
}

filter {
  grok {
    match => {
      "message" => '%{IPORHOST:clientip} %{USER:ident} %{USER:auth} \[%{HTTPDATE:timestamp}\] "%{WORD:verb} %{DATA:request} HTTP/%{NUMBER:httpversion}" %{NUMBER:response:int} (?:-|%{NUMBER:bytes:int}) %{QS:referrer} %{QS:agent}'
    }
  }

  date {
    match => [ "timestamp", "dd/MMM/YYYY:HH:mm:ss Z" ]
    locale => en
  }

  geoip {
    source => "clientip"
  }

  useragent {
    source => "agent"
    target => "useragent"
  }
}

output {
  stdout { codec => dots }
}
```

There's not that many dots(events) between `0.1x` user space GCs imo. 

[gc.txt](https://github.com/elastic/logstash/files/1177530/gc.txt)
